File: matlab/data_associate_known.m
function [zf,idf,zn,table]= data_associate_known(z, idz, table, Nf)
% 

zf= []; zn= [];
idf= []; idn= [];

% find associations (zf) and new features (zn)
for i=1:length(idz)
    ii= idz(i);
    if table(ii) == 0 % new feature
        zn= [zn z(:,i)];
        idn= [idn ii];
    else
        zf= [zf z(:,i)];
        idf= [idf table(ii)];
    end
end

% add new feature IDs to lookup table
table(idn)= Nf + (1:size(zn,2)); % add new feature positions to lookup table

File: testcode/data_associate_known.cpp
#include "data_associate_known.h"
#include <iostream>

void data_associate_known(MatrixXf z, vector<int> idz, VectorXf &table, int Nf, \
        MatrixXf &zf, vector<int> &idf, MatrixXf &zn) 
{
    idf.clear();
    vector<int> idn;

    unsigned i,ii,r;
    
    //this extra loopis required to set the dimension of zn and zf
    int zncols = 0;
    int zfcols = 0;
    for (i =0; i< idz.size(); i++){
        ii = idz[i];
        if (table(ii) ==-1) { //new feature
            zncols++;
        }
        else {
            zfcols++;
        }
    }
   
    //resize
    zn.resize(z.rows(),zncols);
    zn.setZero();
    zf.resize(z.rows(),zfcols);
    zf.setZero();
    
    int znc,zfc;
    znc =0;
    zfc =0;
    for (i =0; i< idz.size(); i++){
        ii = idz[i];
        if (table(ii) ==-1) { //new feature
            for (r=0; r<z.rows();r++) {
                zn(r,znc) = z(r,i);
            }
            znc++;
            idn.push_back(ii);				
        }
        else {
            for (r=0; r<z.rows(); r++) {
                zf(r,zfc) = z(r,i);
            } 
            zfc++;
            idf.push_back(table(ii));
        }	
    }

    for (int i=0; i<idn.size(); i++) {
        table(idn[i]) = Nf+i;  
    }
}

